File: app/src/main/java/za/org/grassroot2/presenter/fragment/MeetingDatePresenter.java
package za.org.grassroot2.presenter.fragment;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import za.org.grassroot2.services.NetworkService;
import za.org.grassroot2.view.FragmentView;

/**
 * Created by qbasso on 21.09.2017.
 */

public class MeetingDatePresenter extends BaseFragmentPresenter<MeetingDatePresenter.MeetingDateView> {

    private final NetworkService networkService;

    @Inject
    MeetingDatePresenter(NetworkService networkService) {
        this.networkService = networkService;
    }

    @Override
    public void onViewCreated() {
        disposableOnDetach(getView().dateInputConfirmed().observeOn(AndroidSchedulers.mainThread()).flatMap(s -> {
            return networkService.getTimestampForText(s);
        }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe(timestamp -> getView().renderDate(timestamp), throwable -> {
            throwable.printStackTrace();
            getView().showDatePicker();
        }));
    }

    public interface MeetingDateView extends FragmentView {
        Observable<String> dateInputConfirmed();
        void renderDate(Long timestamp);
        void showDatePicker();
    }
}

File: app/src/main/java/za/org/grassroot2/presenter/fragment/MeetingDatePresenter.kt
package za.org.grassroot2.presenter.fragment

import javax.inject.Inject

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import za.org.grassroot2.services.NetworkService
import za.org.grassroot2.view.FragmentView

/**
 * Created by qbasso on 21.09.2017.
 */

class MeetingDatePresenter @Inject
internal constructor(private val networkService: NetworkService) : BaseFragmentPresenter<MeetingDatePresenter.MeetingDateView>() {

    override fun onViewCreated() {
        disposableOnDetach(view.dateInputConfirmed().observeOn(AndroidSchedulers.mainThread())
                .flatMap { s -> networkService.getTimestampForText(s) }.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe({ timestamp -> view.renderDate(timestamp) }, { throwable ->
                    throwable.printStackTrace()
                    view.showDatePicker()
                }))
    }

    interface MeetingDateView : FragmentView {
        fun dateInputConfirmed(): Observable<String>
        fun renderDate(timestamp: Long?)
        fun showDatePicker()
    }
}
